[inductor] heuristics based on kernel templates

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156477
* #156292
* __->__ #156282
* #156281
* #156280
* #156279
* #156278
* #156277
* #156276

Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

coconutruben committed:
[inductor] heuristics based on kernel templates

Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

NikhilAPatel reviewed (approved):
LGTM

masnesral reviewed (approved):
Looks fine to me (I'm not sure confident I grok everything that's happening here tho :/ )

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] heuristics based on kernel templates"


Summary:
\# Why

- kernel templates know what kind of parameters they need
- provides an interface to unify other backends behind choices.py as the kernel template can now determine which configheuristics are used

\# What

- pass KernelTemplate to get the config heuristics in addition to the KernelInputs
- use this to unify around `get_mm_config_params` which handles everything internally e.g. TMA is handled by using a different config heuristic as it's a different kernel

To make this work I've started the work of migrating out the things that are triton(mm) specific into the triton_templates e.g.
- the grid itself
- the symbolic grid fn
those are migrated as *copies* for now and will be deleted once they are no longer in use elsewhere

\# Context

https://docs.google.com/document/d/1UoTRm9fWvyUsuft_tN5hUl38Q6n61xAESboMdlspxnw/edit?tab=t.0#heading=h.ddkzcihs3d6l for the detailed technical implementation plan

Test Plan:
```
python3 /data/users/coconutruben/fbsource/fbcode/scripts/coconutruben/torchmm/experiment.py
```

showing regular mm falling back to regular configs. This has been downgraded to a debug statement in the latest version now

```
W0616 19:28:15.031000 166871 site-packages/torch/_inductor/choices.py:137] [0/0] Template ID TritonTemplate::mm not found in template-specific heuristics, using default heuristics
```

showing persistent tma being picked up correctly

```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0719 ms 100.0%
  triton_mm_persistent_tma_19 0.0738 ms 97.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_EXPERIMENTAL_API=True, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8
  triton_mm_17 0.0769 ms 93.4% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
